src/main.cpp

```diff
@@ -7530,30 +7530,35 @@ bool SendMessages(CNode *pto)
         //
         // Message: getdata (blocks)
         //
-        std::vector<CInv> vGetData;
         if (!pto->fDisconnect && !pto->fClient && state.nBlocksInFlight < (int)MAX_BLOCKS_IN_TRANSIT_PER_PEER)
         {
             std::vector<CBlockIndex *> vToDownload;
             FindNextBlocksToDownload(pto->GetId(), MAX_BLOCKS_IN_TRANSIT_PER_PEER - state.nBlocksInFlight, vToDownload);
             // LOG(REQ, "IBD AskFor %d blocks from peer=%s\n", vToDownload.size(), pto->GetLogName());
+            std::vector<CInv> vGetBlocks;
             for (CBlockIndex *pindex : vToDownload)
             {
                 CInv inv(MSG_BLOCK, pindex->GetBlockHash());
                 if (!AlreadyHave(inv))
                 {
-                    if (!IsInitialBlockDownload())
-                        requester.AskFor(inv, pto);
-                    else
-                        requester.AskForDuringIBD(inv, pto);
+                    vGetBlocks.emplace_back(inv);
                     // LOG(REQ, "AskFor block %s (%d) peer=%s\n", pindex->GetBlockHash().ToString(),
-                    //  pindex->nHeight, pto->GetLogName());
+                    //     pindex->nHeight, pto->GetLogName());
                 }
             }
+            if (!vGetBlocks.empty())
+            {
+                if (!IsInitialBlockDownload())
+                    requester.AskFor(vGetBlocks, pto);
+                else
+                    requester.AskForDuringIBD(vGetBlocks, pto);
```

Comment:
wonder if we could have only one `AskFor` to which we pass another bool par to indicate if we are in the IBD phase or not

Reply:
@sickpig I made your requested changed but I've since changed it back.  While in general you are right to suggest this, in this particular case it doesn't work very well.  The problem is that the AskFor() method is also used for getting transactions also; this then makes the code somewhat confusing when I inject code specific for IBD, but in addition we have to then make AskFor() recursive.  I found after I wrote the fix you requested that the code would be very hard to read and make sense for someone unfamiliar with the code, of and coupled with a recursive call that could potentially be a problem if someone in the future updated the code and made an unintended change.  So I thought better to make the code readable and more safe and leave it as it is with the AskForDuringIBD().

Reply:
fair enough